Dashboards api (#164)
metricly/metricly-docs: content/api/api-dashboards.md

```
{
  "dashboards": [
    {
      "id": "df1b9000-1234-4d2a-b0a7-e74e4c5d00c6",
      "userId": 23456,
      "name": "ExampleDashboard",
      "description": null,
      "layout": null,
      "creatorEmail": "support-testing@netuitive.com",
      "created": "2018-05-24T12:59:34Z",
      "updated": "2019-09-19T14:24:49Z",
      "widgets": [],
      "properties": {
        "costOnly": "false",
        "timeRangeDuration": "3600",
        "gridstackContents": "[{\"id\":\"14b640c1-74fb-491a-83a9-e4dbf7cab27c\",\"x\":0,\"y\":0,\"width\":12,\"height\":6}]",
        "refreshIntervalSeconds": "300",
        "globalFilter": "false",
        "wrap": "true"
      },
      "type": "DEFAULT",
      "private": false
    }
  ]
}
```

{{% /expand %}}

---
## POST to /dashboards

{{< button href="https://app.metricly.com/swagger-ui.html#!/dashboards/createDashboardUsingPOST" theme="success" >}} POST {{< /button >}} Use this endpoint to create a dashboard.

{{% expand "View Method Details." %}}

### Parameters

| Parameter | Parameter Type | Data Type | Description |
|--------------------|----------------|-----------|----------------------------------------------|

> | User-Agent | header | string | User agent. |

Reviewer: ```suggestion
| User-Agent | header | string | User-Agent |
```

           "excludedElementScopeFqns": "[]", \
           "grouping": "attribute=SERVICE", \
           "useAllMetricScopeTags": "true", \
           "metricAggs": "[]", \
           "elementScopeIds": "[]", \
           "metricScopeTags": "[]", \
           "groupByPolicy": "false", \
           "useAllElementScopeAttributes": "true", \
           "metricAgg": "avg", \
           "metrics": "[{\"fqn\":\"aws.ec2.cpuutilization\",\"useRegex\":false,\"aggFns\":[\"avg\"],\"aggFn\":null,\"groupAggFn\":null,\"aggregationGroups\":[]}]", \
           "topNLimit": "5", \
           "elementScopeExcludedTags": "[]" \
         }, \
         "generated": false \
       } \
     ], \
     "properties": { \
       "timeRangeDuration": "3600", \
       "costOnly": "false", \
       "gridstackContents": "[{\"id\":\"1\",\"x\":0,\"y\":0,\"width\":12,\"height\":9},{\"id\":\"2\",\"x\":0,\"y\":9,\"width\":12,\"height\":10}]", \
       "refreshIntervalSeconds": "300", \
       "globalFilter": "false", \
       "wrap": "true" \
     }, \
     "type": "DEFAULT", \
     "private": true \
   } \
 }' 'https://app.metricly.com/dashboards'
```

### Swagger Payload

> Use the following payload to test this POST method on Swagger.

Reviewer: ```suggestion
Use the following payload to test this POST method in Swagger.
```

          "grouping": "attribute=SERVICE",
          "useAllMetricScopeTags": "true",
          "metricAggs": "[]",
          "elementScopeIds": "[]",
          "metricScopeTags": "[]",
          "groupByPolicy": "false",
          "useAllElementScopeAttributes": "true",
          "metricAgg": "avg",
          "metrics": "[{\"fqn\":\"aws.ec2.cpuutilization\",\"useRegex\":false,\"aggFns\":[\"avg\"],\"aggFn\":null,\"groupAggFn\":null,\"aggregationGroups\":[]}]",
          "topNLimit": "5",
          "elementScopeExcludedTags": "[]"
        },
        "generated": false
      }
    ],
    "properties": {
      "timeRangeDuration": "3600",
      "costOnly": "false",
      "gridstackContents": "[ {\n  \"id\" : \"8c7e11c0-b88c-4393-91a6-20ca053935ec\",\n  \"x\" : 0,\n  \"y\" : 0,\n  \"width\" : 12,\n  \"height\" : 9\n}, {\n  \"id\" : \"eb98c261-a3dd-4961-9628-fccc191fb846\",\n  \"x\" : 0,\n  \"y\" : 9,\n  \"width\" : 12,\n  \"height\" : 10\n} ]",
      "refreshIntervalSeconds": "300",
      "globalFilter": "false",
      "wrap": "true"
    },
    "type": "DEFAULT",
    "private": true
  }
}
```

{{% /expand %}}

---

## GET from /dashboards/dashboardtype/{type}

> {{< button href="https://app.metricly.com/swagger-ui.html#!/dashboards/getByDashboardTypeUsingGET" theme="info" >}} GET {{< /button >}} Use this endpoint to get packaged dashboards specific to certain element types.

Reviewer: ```suggestion
{{< button href="https://app.metricly.com/swagger-ui.html#!/dashboards/getByDashboardTypeUsingGET" theme="info" >}} GET {{< /button >}} Use this endpoint to get dashboards specific to a certain element type.
```

      },
      "type": "EC2",
      "private": false
    },
    {
      "id": "S3 BUCKET",
      "userId": null,
      "name": "AWS S3 Element Detail Page",
      "description": null,
      "layout": null,
      "creatorEmail": "research@netuitive.com",
      "created": "2018-09-20T21:49:28Z",
      "updated": "2018-09-20T21:49:28Z",
      "widgets": [],
      "properties": {
        "integrationType": null,
        "density": "normal",
        "timeRangeDuration": "3600",
        "gridstackContents": "[ {\n  \"id\" : \"9d1ad310-a2e2-3ddd-819d-4ef9ca1c85f7\",\n  \"x\" : 0,\n  \"y\" : 9,\n  \"width\" : 12,\n  \"height\" : 9\n}, {\n  \"id\" : \"44775f23-5589-3b16-a7c2-ace703563875\",\n  \"x\" : 0,\n  \"y\" : 0,\n  \"width\" : 12,\n  \"height\" : 9\n} ]",
        "refreshIntervalSeconds": "300",
        "wrap": "true"
      },
      "type": "S3 BUCKET",
      "private": false
    }
  ]
}
```

{{% /expand %}}

---

> ## DELETE from /dashboards/{id}

Reviewer: ```suggestion
## DELETE to /dashboards/{id}
```

{{< button href="https://app.metricly.com/swagger-ui.html#!/dashboards/deleteDashboardUsingDELETE" theme="danger" >}} DELETE {{< /button >}} Use this endpoint to delete a dashboard.

{{% expand "View Method Details." %}}

### Parameters

| Parameter | Parameter Type | Data Type | Description |
|--------------------|----------------|-----------|----------------------------------------------|

> | User-Agent  | header  | string  | User agent.   |

Reviewer: ```suggestion
| User-Agent  | header  | string  | User-Agent  |
```

```

### Response Body

The following response body returns basic details and properties of the dashboard.

```
{
  "dashboard": {
    "id": "23d3a5e6-0000-1111-9999-d5abe01ff59c",
    "userId": 87654,
    "name": "Example Name",
    "description": null,
    "layout": null,
    "creatorEmail": "support-testing@netuitive.com",
    "created": "2016-06-29T16:18:47Z",
    "updated": "2019-04-08T20:15:43Z",
    "widgets": [],
    "properties": {
      "integrationType": null,
      "timeRangeDuration": "28800",
      "costOnly": "false",
      "gridstackContents": "[]",
      "refreshIntervalSeconds": "300",
      "globalFilter": "false",
      "wrap": "true"
    },
    "type": "DEFAULT",
    "private": false
  }
}

```

{{% /expand %}}

---
## PUT to /dashboards/{id}

{{< button href="https://app.metricly.com/swagger-ui.html#!/dashboards/getSingleUsingGET" theme="warning" >}} PUT {{< /button >}} Use this endpoint to update a dashboard's settings.

{{% expand "View Method Details." %}}

### Parameters

| Parameter | Parameter Type | Data Type | Description |
|--------------------|----------------|-----------|----------------------------------------------|

> | User-Agent  | header  | string  | User agent.  |

Reviewer: ```suggestion
| User-Agent  | header  | string  | User-Agent  |
```

          "groupByPolicy": "false",
          "useAllElementScopeAttributes": "true",
          "metricAgg": "avg",
          "metrics": "[]",
          "topNLimit": "5",
          "elementScopeExcludedTags": "[]"
        },
        "generated": false
      }
    ],
    "properties": {
      "timeRangeDuration": "3600",
      "costOnly": "false",
      "gridstackContents": "[{\"id\":\"c63f2aad-7167-45fd-97cf-149ffe29ba6a\",\"width\":4,\"height\":7,\"x\":0,\"y\":0}]",
      "refreshIntervalSeconds": "300",
      "globalFilter": "false",
      "wrap": "true"
    },
    "type": "DEFAULT",
    "private": false
  }
}
```

### Response Body

The following response body returns no content and a 204 success code.

```
no content
```

{{% /expand %}}

---

## POST to /dashboards/{id}/copy

{{< button href="https://app.metricly.com/swagger-ui.html#!/dashboards/copyDashboardUsingPOST" theme="success" >}} POST {{< /button >}} Use this endpoint to copy a dashboard.

{{% expand "View Method Details." %}}

### Parameters

| Parameter | Parameter Type | Data Type | Description |
|--------------------|----------------|-----------|----------------------------------------------|

> | User-Agent   | header  | string   | User agent.  |

Reviewer: ```suggestion
| User-Agent   | header  | string   | User-Agent |
```

```
![grid-stack-example](/images/dashboards/grid-stack-example.png)

{{% /expand %}}

## Create a New Dashboard Based on an Existing Dashboard

> There are two ways you can quickly create a dashboard: clone (POST) and update (PUT), or GET and POST. This example uses the GET and POST endpoints.

Reviewer: ```suggestion
There are two ways to build a dashboard: copy an existing dashboard (POST) and make an update to it (PUT), or retrieve an existing dashboard (GET) and use it as a starting point to create a new one (POST). This example uses the GET and POST endpoints.
```
